src/annotator/highlighter.ts

```diff
@@ -385,3 +393,146 @@ export function getBoundingClientRect(collection: HTMLElement[]): Rect {
     right: Math.max(acc.right, r.right),
   }));
 }
+
+/**
+ * Add metadata and manipulate ordering of all highlights in `element` to
+ * allow styling of nested, clustered highlights.
+ */
+export function updateClusters(element: Element) {
+  setNestingData(getHighlights(element));
+  updateSVGHighlightOrdering(element);
+}
+
+/**
+ * Is `el` a highlight element? Work around inconsistency between HTML documents
+ * (`tagName` is upper-case) and XHTML documents (`tagName` is lower-case)
+ */
+const isHighlightElement = (el: Element): boolean =>
+  el.tagName.toLowerCase() === 'hypothesis-highlight';
+
+/**
+ * Return the closest generation of HighlightElements to `element`.
+ *
+ * If `element` is itself a HighlightElement, return immediate children that
+ * are also HighlightElements.
+ *
+ * Otherwise, return all HighlightElements that have no parent HighlightElement,
+ * i.e. the outermost highlights within `element`.
+ */
+function getHighlights(element: Element) {
+  let highlights;
+  if (isHighlightElement(element)) {
+    highlights = Array.from(element.children).filter(isHighlightElement);
+  } else {
+    highlights = Array.from(
+      element.getElementsByTagName('hypothesis-highlight')
+    ).filter(
+      highlight =>
+        !highlight.parentElement || !isHighlightElement(highlight.parentElement)
```

Comment:
Given that we are searching for children of `element` here, `parentElement` should always be non-null and thus `highlight => !isHighlightElement(highlight.parentElement!)` should work.